### PR TITLE
arch: arc: Add mpu support

### PR DESCRIPTION
**Pull request by vonhust**
* add arc mpu driver
* modify the corresponding kconfig and kbuild
* currently only em_starterkit 2.2's em7d configuration
  has mpu feature (mpu version 2)
* as the minimum region size of arc mpu version 2 is 2048 bytes and
  region size should be power of 2, the stack size of threads
  (including main thread and idle thread) should be at least
  2048 bytes and power of 2
* for mpu stack guard feature, a stack guard region of 2048 bytes
  is generated. This brings more memory footprint
* For arc mpu version 3, the minimum region size is 32 bytes.
* the codes are tested by the mpu_stack_guard_test and stackprot

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

### CODE REVIEW

File: arch/arc/core/mpu/arc_mpu.c [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+ * type (i.e. THREAD_STACK_REGION) and return the correct parameter set.
+ */
+static inline u32_t _get_region_attr_by_type(u32_t type, u32_t size)
+{
+	switch (type) {
+	case THREAD_STACK_REGION:
+		return 0;
+	case THREAD_STACK_GUARD_REGION:
+	/* no RW to guard region */
+		return  AUX_MPU_RDP_REGION_SIZE(
```
Comment (chunlinhan) [inline arch/arc/core/mpu/arc_mpu.c]: should be
return AUX_MPU_RDP_REGION_SIZE(ARC_FEATURE_MPU_ALIGNMENT_BITS);
or
return AUX_MPU_RDP_REGION_SIZE(_vmpu_region_bits(1 << ARC_FEATURE_MPU_ALIGNMENT_BITS));
?

Comment (vonhust) [inline arch/arc/core/mpu/arc_mpu.c]: Thanks @chunlinhan . It's an error, should be AUX_MPU_RDP_REGION_SIZE(ARC_FEATURE_MPU_ALIGNMENT_BITS);.

I will modify it in the coming commits

File: arch/arc/core/mpu/arc_mpu.c [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	u32_t i;
+
+	num_regions = _get_num_regions();
+
+	/* ARC MPU supports up to 16 Regions */
+	if (mpu_config.num_regions > num_regions) {
+		return;
+	}
+
+	r_index = num_regions - mpu_config.num_regions;
```
Comment (ruuddw) [inline arch/arc/core/mpu/arc_mpu.c]: Please add some explanation here: on use of regions: first 'special type' regions, then static ones from mpu_config struct.

File: arch/arc/core/mpu/arc_mpu.c [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	/* Disable MPU */
+	arc_core_mpu_disable();
+
+	for (i = 0; i < r_index; i++) {
+		_region_init(i, 0, 0);
+	}
+
+	/* Configure regions */
+	for (r_index = 0; i < num_regions; i++) {
```
Comment (ruuddw) [inline arch/arc/core/mpu/arc_mpu.c]: add a note that loop over i continues with remaining static items.

File: arch/arc/core/mpu/arc_mpu.c [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		_region_init(i,
+			mpu_config.mpu_regions[r_index].base
+			| AUX_MPU_RDB_VALID_MASK,
+			mpu_config.mpu_regions[r_index].attr);
+		r_index++;
+	}
+
+	/* Enable MPU */
```
Comment (ruuddw) [inline arch/arc/core/mpu/arc_mpu.c]: I think the default region should be updated here as well (disabled, so addresses not explicitly defined in MPU no longer have the 'everything permitted' reset setting.

File: arch/arc/core/mpu/arc_mpu.c [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	 *
+	 * For ARC MPU v2, MPU regions can be overlapped, smaller
+	 * region index has higher priority.
+	 */
+	u32_t region_index = _get_num_regions() - mpu_config.num_regions
+			- type;
+	u32_t region_attr = _get_region_attr_by_type(type, size);
+
+	/* ARC MPU supports up to 16 Regions */
+	if (region_index == 0) {
```
Comment (ruuddw) [inline arch/arc/core/mpu/arc_mpu.c]: Don't think this check will work, only catches case where total number of regions equals mpu_config regions + 'type' region. Please change into a better check (type < _get_nu,_regions() - mpu_config.num_regions, or something like that)

**Review by ruuddw — CHANGES_REQUESTED** [review]
Initial MPU support for ARC. Looks mostly ok, see detailed remarks for few improvement suggestions.

File: arch/arc/core/mpu/Kconfig [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	default n
+	help
+	  ARC Core MPU functionalities
+
+config MPU_STACK_GUARD
+	bool "Thread Stack Guards"
+	depends on ARC_CORE_MPU
+	default n
+	help
+	  Enable Thread Stack Guards via MPU
```
Comment (andrewboie) [inline arch/arc/core/mpu/Kconfig]: There are *much* better ways of doing stack checking on ARC, see CONFIG_ARC_STACK_CHECKING.

We need to clearly document that CONFIG_MPU_STACK_GUARD is very, very, sub-optimal configuration and should only be used on those platforms that don't support CONFIG_ARC_STACK_CHECKING.

File: include/arch/arc/arch.h [hunk 7]
```diff
@@ -38,7 +38,107 @@ extern "C" {
...
+#endif
+#else /* CONFIG_ARC_CORE_MPU */
+#error "Unsupported STACK_ALIGN"
+#endif
+#else  /* CONFIG_MPU_STACK_GUARD */
 #define STACK_ALIGN  4
+#endif
+
+/**
+ * @brief Declare a toplevel thread stack memory region
```
Comment (andrewboie) [inline include/arch/arc/arch.h]: these APIS are already documented in kernel.h, please don't copy-paste the documentation here

File: arch/arc/core/mpu/Kconfig (hunk 6 above)
Comment (vonhust) [inline arch/arc/core/mpu/Kconfig]: For stack guard,  ARC has more effective way,  i.e., STACK_CHECKING.  But CONFIG_MPU_STACK_GUARD is added here to test the MPU feature.  After I search the zephyr's codes,  it seems MPU_STACK_GUARD is the only place to show MPU's protection.  In the future, it's better to use STACK_CHECKING to do stack protection.

**Comment by vonhust** [comment]
The configure of default mpu region is done in arc_core_mpu_enable().
No permissions are given to default mpu region.

File: arch/arc/core/mpu/arc_mpu.c (hunk 4 above)
Comment (vonhust) [inline arch/arc/core/mpu/arc_mpu.c]: The default regions is configured in arc_core_mpu_enable

**Review by ruuddw — APPROVED** [review]

**Pull request merged**